File: LeetCode/Medium/0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ret = 0;
    
    int distributeCoins(TreeNode* root) {
        dfs(root);
        
        return ret;
    }
    
    int dfs(TreeNode* node){
        if (!node) return 0;
        
        node->val += dfs(node->left) + dfs(node->right);
        ret += abs(node->val - 1);
        
        return node->val - 1;
    }
};

File: LeetCode/Medium/0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def distributeCoins(self, root: Optional[TreeNode]) -> int:
        ret = [0]
        def dfs(node):
            if not node:
                return 0
            
            l, r = dfs(node.left), dfs(node.right)
            node.val += l + r

            ret[0] += abs(node.val - 1)

            return node.val - 1

        dfs(root)

        return ret[0]